pkg/utils/pod/scheduling.go

```diff
@@ -45,6 +45,29 @@ func IsSchedulable(pod *v1.PodSpec, node *v1.Node) bool {
 	return true
 }
 
+// IgnoredForUnderutilization returns true if the set of pods has no non-daemonset pods
+func IgnoredForUnderutilization(pods []*v1.Pod) bool {
+	for _, p := range pods {
+		if HasFailed(p) {
+			continue
+		}
+		if !IsOwnedByDaemonSet(p) {
+			return false
+		}
+	}
+	return true
+}
+
+// AreAllTerminating returns true if the set of pods were all unable to join
+func AreAllTerminating(pods []*v1.Pod) bool {
+	for _, p := range pods {
```

Comment:
I think you want to reuse your `IgnoredForUnderutilization` helper, since this code should also take into account `HasFailed`. It does make the naming get a bit weird, so maybe we should just check HasFailed here.

I think there must be a way to factor this to share the concept. It is the same concept after all, we just need a name for it.

Reply:
These are a little different. We decided that we only wanted to terminate the node if all of the pods were terminating. This means we need to check the deletionTimestamp, whereas the ignoredForUnderutilization also checks if they failed, which doesn't necessarily mean they were requested to be deleted yet.

Reply:
What if a pod is failed (i.e. scheduling misbehavior)? I think we should still terminate the node in this case. This is for the exact same reason you want to consider failed pods in the other function. 

This similarity leads me to believe that there is common behavior that should be factored out.